### ISSUE TITLE
Applying decorators conditionally using type constraints

### ISSUE THREAD

**Issue by meinsiedler** [khellang/Scrutor#159]
Hi!

We are using Scrutor to apply generic decorators to command handlers and query handlers. This means we have for example a generic `ICommandHandler<TCommand>` interface and a specific handler implementation which then handles a specific command, e.g. `AddPersonCommandHandler : ICommandHandler<AddPerson>`.

We apply generic decorators for cross-cutting concerns like transactions, logging, and so on.

Also, we use decorators for permission checks, this looks like this for example:

```csharp
class AccessValidationCommandHandlerDecorator<TCommand> : ICommandHandler<TCommand>
    where TCommand : IAccessRestricted
{
    private readonly ICommandHandler<TCommand> decoratee;

    public AccessValidationCommandHandlerDecorator(ICommandHandler<TCommand> decoratee)
    {
        this.decoratee = decoratee;
    }

    void ICommandHandler<TCommand>.Handle(TCommand command)
    {
        // Do access validation
        this.decoratee.Handle(command);
    }
}
```

The decorator in this example uses a generic type constraint for `TCommand` so that the decorator should only apply to those `ICommandHandler<TCommand>` implementations where the `TCommand` implements the `IAccessRestricted` interface.

We are coming from Simple Injector where this use case is supported and documented here: https://docs.simpleinjector.org/en/latest/aop.html#applying-decorators-conditionally-using-type-constraints

It seems like Scrutor doesn't consider the generic type constraints and tries to apply the decorator to all `ICommandHandler<TCommand>` implementations regardless of the generic type constraint which leads to an exception on startup because the generic type constraint is not satisfied.

Would it be possible to support that use case and that Scrutor considers the generic type constraints when applying decorators?

**Comment by khellang** [comment]
> Would it be possible to support that use case and that Scrutor considers the generic type constraints when applying decorators?

It would, yes. I don't think this would be breaking (given it's implementation is sound) as decoration today simply breaks if applied to types with incompatible constraints. It's not trivial to implement though. I probably won't have time to do this in the short term, but if someone were to provide a PR with this feature, I'd be glad to review and merge it 😄

**Comment by khellang** [comment]
I wonder if this case was actually "solved" by #139? It's basically just catching the exception, which has been "best practice" (instead of trying to re-implement the CLR checks) for a while now.

**Comment by meinsiedler** [comment]
Hi,

thank you for your responses. Yes, #139 indeed resolves this issue. I just tested with the new Scrutor version 4.0.0.

So, this issue can be closed. 😊
